src/osd/PrimaryLogPG.cc

```diff
@@ -6844,8 +6954,14 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 			   CEPH_OSD_COPY_FROM_FLAG_MAP_SNAP_CLONE |
 			   CEPH_OSD_COPY_FROM_FLAG_RWORDERED;
 	  unsigned src_fadvise_flags = LIBRADOS_OP_FLAG_FADVISE_SEQUENTIAL;
-	  start_copy(cb, ctx->obc, src_hoid, my_oloc, 0, flags,
-		     obs.oi.soid.snap == CEPH_NOSNAP,
+	  ObjectContextRef src_obc = get_object_context(src_hoid, false, NULL);
```

Comment:
What is src_hoid here?  I thought it must be the redirect target for a redirect or the various chunk targets for a chunked manifest, but it appears to be local to this pg?

Reply:
Oh, it works because in either case the start_copy machinery will send a read to this very osd which will cause it to be proxied to the correct set of targets.  That seems kind of brittle.

Given that, I'm not sure I understand why you are setting src_hoid.snap to ctx->snapc.seq?  If you're trying to promote a clone, that seems wrong.

Reply:
Can you promote a clone at all?

Reply:
In chunked manifest implementation, metadata (object_info_t...) is stored upper tier and data (the content of chunks) is stored lower tier. So, the pg in upper tier has manifest.

Reply:
I'm not following how that's related to src_hoid.snap.

Reply:
Snapshots have different chunk_maps. For example, the head has chunk_map <1, 2> older clone has chunk_map <1,3>.  So, to read specific version of snapshot, this chunk_map is needed.
To do that, get_object_context will read the chunk_map in object_info_t based on src_hoid.snap.  After then, start_copy will read chunks based on referring to the chunk_map

Reply:
I suppose it doesn't matter as no other write can occur while its happening.

Reply:
Please take a look the following test case.

```
ioctx.selfmanaged_snap_remove(my_snaps[1]);
// rollback snap
ioctx.selfmanaged_snap_rollback("foo", my_snaps[0]);
vector<uint64_t> my_snaps_promote(1);
my_snaps_promote[0] = my_snaps[0];
ASSERT_EQ(0, ioctx.selfmanaged_snap_set_write_ctx(my_snaps_promote[0], my_snaps_promote));

// promote snap
{
    ObjectWriteOperation op;
    op.tier_promote();
}
```

A client rollback snap and then, it invokes tier_promote() to read rollbacked snaphost.
At this time, selfmanaged_snap_set_write_ctx is called to set ctx's seq.
This commit is added to handle the above case.
Am I wrong ? or does this case never happens?

Reply:
Oh, that might actually be wrong.  A rollback is a write.  So let's say that we perform the following operations:

writefull "foo" snapc=1:[1]
writefull "bar" snapc=2:[2,1]
writefull "baz" snapc=3:[3,2,1]
rollback 2 snapc=4:[4,3,2,1]

read snap=1  -> ENOENT
read snap=2 -> "foo"
read snap=3 -> "bar"
read snap=4 -> "baz"
read snap=HEAD -> "bar"

I'm not sure what the purpose of the

ASSERT_EQ(0, ioctx.selfmanaged_snap_set_write_ctx(my_snaps_promote[0], my_snaps_promote));

line is.  Is the intent to remove the end of that vector to remove snapshots?  In either case, if the osd's recorded snapc for the object has a more recent seq than the one you send, it just ignores yours (see PrimaryLogPG::make_writeable).

I would have expected tier_promote to have no impact on future reads, so I'd assume it always reads from the current head version of the object.

Was there a previous conversation I can look at to fill me in on the background?

Reply:
There is no discussion about above case. I just made a test case by guessing how to use after looking at the current API (selfmanaged_snap_*). If the exact intent of snap is what you've mentioned, it would be better to erase this part following that (related to src_hoid.snap).